File: frontend/src/app/models/notifications/notification.ts
export class Notification {
  #id!: string;
  #userId!: string;
  #postId!: string;
  #username!: string;
  #userProfile!: string;
  #title!: string;
  #message!: string;
  #createdDate!: Date;

  get id(): string {
    return this.#id;
  }

  get userId(): string {
    return this.#userId;
  }

  get postId(): string {
    return this.#postId;
  }

  get username(): string {
    return this.#username;
  }

  get userProfile(): string {
    return this.#userProfile;
  }

  get title(): string {
    return this.#title;
  }

  get message(): string {
    return this.#message;
  }

  get createdDate(): Date {
    return this.#createdDate;
  }

  set id(value: string) {
    this.#id = value;
  }

  set userId(value: string) {
    this.#userId = value;
  }

  set postId(value: string) {
    this.#postId = value;
  }

  set username(value: string) {
    this.#username = value;
  }

  set userProfile(value: string) {
    this.#userProfile = value;
  }

  set title(value: string) {
    this.#title = value;
  }

  set message(value: string) {
    this.#message = value;
  }

  set createdDate(value: Date) {
    this.#createdDate = value;
  }

  static fromJSON(json: any): Notification {
    const notification = new Notification();
    notification.id = json.id;
    notification.userId = json.userId;
    notification.postId = json.postId;
    notification.username = json.username;
    notification.userProfile = json.userProfile;
    notification.title = json.title;
    notification.message = json.message;
    notification.createdDate = new Date(json.createdDate);
    return notification;
  }
}

File: backend/Domain/Notifications/Notification.cs
using Backend.Domain.Primitive;
using Backend.Domain.Users;
using Backend.Domain.Enums;
using Backend.Domain.Posts;
using MongoDB.Bson.Serialization.Attributes;

namespace Backend.Domain.Notifications;

public class Notification : Entity
{
    private const string DEFAULT_TITLLE = "commented on your post";
    public NotificationId Id { get; private set; } = default!;
    public UserId UserId { get; private set; } = default!;
    public PostId PostId { get; private set; } = default!;
    [BsonIgnore]
    public User? User { get; private set; }
    public string Title { get; private set; } = DEFAULT_TITLLE;
    public string Message { get; private set; } = string.Empty;
    public Status Status { get; private set; }
    public DateTime CreatedDate { get; private set; }
    public DateTime? DeletedDate { get; private set; }

    private Notification() { }

    private Notification(UserId userId, PostId postId, string message)
    {
        Id = new NotificationId(Guid.NewGuid());
        UserId = userId;
        PostId = postId;
        Message = message;
        Status = Status.Active;
        CreatedDate = DateTime.UtcNow;
    }

    public void SetUser(User user){
        User = user;
    }

    public static Notification Create(UserId userId, PostId postId, string message)
    {
        if (userId is null)
            throw new ArgumentNullException(nameof(userId));
        if (string.IsNullOrWhiteSpace(message))
            throw new InvalidOperationException("Message cannot be empty.");
        var notification = new Notification(userId, postId, message);
        return notification;
    }
}